### PR TITLE
YaccParserErrorKind::DuplicateRule add an original_span field

### PR DESCRIPTION
**Pull request by ratmice**
While YaccParserError has the span of the duplicate, original_span contains a span for the rule being duplicated.
Issue #300 task #2 (SemVer incompatible) extra spans, has quite a few variants which could potentially use extra spans.

This patch is kind of an exploratory commit which implements it just for `YaccParserErrorKind::DuplicateRule`, 
so that we could review it and future patches for other `Duplicate*` error kinds, could perhaps follow a similar structure.

There is one place in the doc comment specific to the variant it refers to a *rule*, and one it is specific to the error which contains the kind. I couldn't think of any good variant neutral word, and figured that clarity was more important than convenience here.

*note*: I am uncertain if all `Duplicate*` and similar variants have easily accessible spans, for instance `MissingBrace` may not be tracking the original brace.

I need to update my project to master, so I can see how these end up looking in context...

### CODE REVIEW

File: cfgrammar/src/lib/yacc/parser.rs [hunk 1]
```diff
@@ -22,7 +22,12 @@ pub enum YaccParserErrorKind {
     IllegalName,
     IllegalString,
     IncompleteRule,
-    DuplicateRule,
+
+    DuplicateRule {
+        /// While [`YaccParserError`] has the span of the duplicate.<br/>
+        /// `original_span` contains a span for the rule being duplicated.
+        original_span: Span,
```
Comment (ratmice) [inline cfgrammar/src/lib/yacc/parser.rs:29]: I guess now, I wonder if perhaps this would be better as something like `original: DuplicatedItem{DuplicateRule, span}`.
which could allow us to implement an `Error` like `format`, and message?

Comment (ltratt) [reply cfgrammar/src/lib/yacc/parser.rs:29]: I'm not going to pretend I've followed that!

I'm broadly supportive of the general idea behind this PR *but* if we're going to break the API, let's make it more future proof. It should cope with multiple copies, so something like `DuplicateRules(Vec<Span>)` would be safer in the long run.

Comment (ratmice) [reply cfgrammar/src/lib/yacc/parser.rs:29]: Sorry about that, was attempting to sleep and write stuff down at the same time.

The general gist of it is, I was wondering if we want to provide a string message along with the span,
that mirrors the original "Duplicate rule" string but provides some context for the span.

like "The first instance of the duplicated rule", "additional duplicate rule",
we'll need some enum to match on, for implementing format strings, and it might make sense to just use the `newtype` pattern and use `YaccParserErrorKind` as exists in master for that.

So we could have something like `Vec<Newtype(YaccParserErrorKind, Span)>` in error where `newtype` provides:

```
// in the newtype
match self.kind {
   YaccParserErrorKind::DuplicateRule = f("the original rule duplicated"),
   _ => { "Does not have additional data" },
}
```

The `Other` variant with `YaccParserErrorRelatedDataKind`, was for fields which don't exactly correspond to one of the top-level error kinds, like "additional duplicate rule".

So basically I was thinking that perhaps we shouldn't add the span to kind, but put it along side a kind in the `YaccParserError`.

Part of the thinking is that this would be more future proof for errors we can't easily get spans for now, but possibly could
in the future, we can just leave an empty `Vec<(newtype, span)>` when we can't.

Comment (ltratt) [reply cfgrammar/src/lib/yacc/parser.rs:29]: Nope, I still don't understand :)

**Comment by ratmice** [comment]
Going to go mark this as draft, as I would like to play with some designs locally which provide a string context for the additional span.

File: cfgrammar/src/lib/yacc/parser.rs (hunk 1 above)
Comment (ratmice) [reply cfgrammar/src/lib/yacc/parser.rs:29]: No worries, trying to implementing it, will perhaps be better than trying to explain it, and it may end up a terrible idea.

**Comment by ratmice** [comment]
So, i went ahead and pushed a commit for which lays out the general idea.

I kind of want to lay out the thought process behind it:

1. It is much easier to be semver compatible and incremental work towards adding spans, if we don't have to add the span to each `YaccParserErrorKind` variants.  So instead this adds a Vec<(kind, Span)> as a field to the error.
2. It makes it non-exhaustive so we can add additional Specialized errors in the future, in theory you shouldn't need to match on it    Just turn them into strings?

This also changes the `Display` impl of `YaccParserError` around, moving it mostly into `YaccParserErrorKind`,
So that we can reuse it.

It is certainly much less polished than my original commit, but I think it may be a more flexible mechanism in the future?

Anyhow, I'm curious your thoughts on the matter.

**Comment by ltratt** [comment]
I must admit that I think this is too complex, and we're also trying to do two things in one PR, when it would be better to split them in two. I would prefer this PR *only* tries to deal with "here are the >1 place you tried to define rule R". Then a second PR could try to deal with "how do we reflect multiple errors in the grammar?" [An obvious thing to try for the second PR would be to return `Result<..., Vec<ParserError>>` or similar rather than a substruct.]

**Comment by ratmice** [comment]
Sounds like a plan, I'm going to go ahead and close this, since that seems like it won't have much in common with either of the previous two commits.

I agree that it is probably too complex too, i.e. that we only actually use the `Kind` now in the `Display` impl of `ParserError` does seem like it could allow for some simplifications, and that some of the complexity here is probably incidental to the previous design.

So this seems like a good call.
